homeassistant/components/landisgyr_heat_meter/config_flow.py

```diff
@@ -90,48 +99,45 @@ async def validate_and_create_entry(self, dev_path):
             data=data,
         )
 
-    async def validate_ultraheat(self, port: str):
+    async def validate_ultraheat(self, port: str) -> tuple[str, str]:
         """Validate the user input allows us to connect."""
 
-        reader = UltraheatReader(port)
-        heat_meter = HeatMeterService(reader)
+        reader = ultraheat_api.UltraheatReader(port)
+        heat_meter = ultraheat_api.HeatMeterService(reader)
         try:
             async with async_timeout.timeout(ULTRAHEAT_TIMEOUT):
                 # validate and retrieve the model and device number for a unique id
                 data = await self.hass.async_add_executor_job(heat_meter.read)
-                _LOGGER.debug("Got data from Ultraheat API: %s", data)
 
-        except Exception as err:
+        except (asyncio.TimeoutError, serial.serialutil.SerialException) as err:
             _LOGGER.warning("Failed read data from: %s. %s", port, err)
             raise CannotConnect(f"Error communicating with device: {err}") from err
 
-        _LOGGER.debug("Successfully connected to %s", port)
+        _LOGGER.debug("Successfully connected to %s. Got data: %s", port, data)
         return data.model, data.device_number
 
-    async def get_ports(self) -> dict:
-        """Get the available ports."""
-        ports = await self.hass.async_add_executor_job(serial.tools.list_ports.comports)
-        formatted_ports = {}
-        for port in ports:
-            formatted_ports[
-                port.device
-            ] = f"{port}, s/n: {port.serial_number or 'n/a'}" + (
-                f" - {port.manufacturer}" if port.manufacturer else ""
-            )
-        formatted_ports[CONF_MANUAL_PATH] = CONF_MANUAL_PATH
-        return formatted_ports
-
 
-def get_serial_by_id(dev_path: str) -> str:
-    """Return a /dev/serial/by-id match for given device if available."""
-    by_id = "/dev/serial/by-id"
-    if not os.path.isdir(by_id):
-        return dev_path
+async def get_usb_ports(hass: HomeAssistant) -> dict[str, str]:
```

Comment:
This function sounds like it shouldn't belong in a single integration..?

Reply:
Possibly, but that's out of scope for this PR.